drivers/SmartThings/matter-rvc/src/init.lua

```diff
@@ -28,213 +25,232 @@ if version.api < 10 then
   end
 end
 
-local function get_field_labels_of_supported_modes(device, field_prefix)
-  local field = string.format("%s_labels", field_prefix)
-  return device:get_field(field)
+local function info_changed(driver, device, event, args)
+  for cap_id, attributes in pairs(subscribed_attributes) do
+    if device:supports_capability_by_id(cap_id) then
+      for _, attr in ipairs(attributes) do
+        device:add_subscribed_attribute(attr)
+      end
+    end
+  end
+  device:subscribe()
 end
 
-local function get_labels_of_supported_modes(device, field_prefix, cluster_id, attribute_id, response)
-  local supported_modes = get_field_labels_of_supported_modes(device, field_prefix)
-  if supported_modes ~= nil then
-    return supported_modes
+-- Helper functions --
+local function update_supported_arguments(device, current_run_mode, operating_state)
+  device.log.info(string.format("update_supported_arguments: %s, %s", current_run_mode, operating_state))
+  if current_run_mode == nil or operating_state == nil then
+    return
   end
 
-  set_field_supported_modes_from_response(device,
-    field_prefix,
-    cluster_id,
-    attribute_id,
-    response)
-  return get_field_labels_of_supported_modes(device, field_prefix)
-end
-
-local function is_idle_mode(device, field_prefix, index, idle_mode_tag)
-  local mode_tags = get_field_mode_tags_of_supported_mode(device, field_prefix, index)
-  for _, mode_tag in ipairs(mode_tags) do
-    if mode_tag == idle_mode_tag then
-      return true
-    end
+  -- Error state
+  if operating_state == "Error" then
+    -- Set runMode to empty
+    local component = device.profile.components["runMode"]
+    local event = capabilities.mode.supportedArguments({}, {visibility = {displayed = false}})
+    device:emit_component_event(component, event)
+    -- Set cleanMode to empty
+    component = device.profile.components["cleanMode"]
+    device:emit_component_event(component, event)
+    return
   end
-  return false
-end
 
-local function get_labels_of_supported_modes_filter_by_current_mode(device, field_prefix, idle_mode_tag, current_mode)
-  local labels = get_field_labels_of_supported_modes(device, field_prefix)
-  local is_idle_current_mode = false
-  for i, label in ipairs(labels) do
-    if label == current_mode then
-      is_idle_current_mode = is_idle_mode(device, field_prefix, i, idle_mode_tag)
+  -- Get the tag of the current run mode
+  local current_tag = 0xFFFF
+  local supported_run_modes = device:get_field(RUN_MODE_SUPPORTED_MODES) or {}
+  for i, mode in ipairs(supported_run_modes) do
+    if mode[1] == current_run_mode then
+      current_tag = mode[2]
       break
     end
   end
+  if current_tag == 0xFFFF then
+    return
+  end
 
-  if is_idle_current_mode then
-    return labels
+  -- Check whether non-idle mode can be selected or not
+  local op_state = capabilities.robotCleanerOperatingState.operatingState
+  local can_be_selected_as_non_idle = 0
```

Comment:
nit: 
```suggestion
  local can_be_non_idle = false
```
and then true in the other places where that makes sense